### ISSUE TITLE
Encapsulation design

### ISSUE THREAD

**Issue by Peanuuutz** [akuleshov7/ktoml#93]
### Discussed in https://github.com/akuleshov7/ktoml/discussions/92

<div type='discussions-op-text'>

<sup>Originally posted by **Peanuuutz** January 12, 2022</sup>
Currently encapsulation is not present, which is an urgent problem for a widely used application. We should only expose those classes and functions that are necessary for users. This will help users to quickly find what they want, as well as help us develop new features as well as modify existing code without worrying about breaking users' applications.

> I'm going to copy [what I have done](https://github.com/Peanuuutz/tomlkt/tree/master/src/commonMain/kotlin/net/peanuuutz/tomlkt). (Actually these are quite common)

The following are public:
- Entrypoint for serialization and deserialization: `Toml`. (Consider [this](https://github.com/akuleshov7/ktoml/issues/49#issuecomment-1003652946) suggestion)
- AST: `TomlValue`(which derives `TomlBoolean`, `TomlLong`, `TomlBasicString`, `TomlArray`, `TomlTable` to name a few).
- AST manipulation: like `JsonPrimitive.int`.
- Exceptions: only sealed `TomlEncodingException` and `TomlDecodingException`.

The following are internal:
- Encoders
- Emitters
- Decoders
- Parsers

Also, I suggest separating AST itself and the actual parsing process, like `Toml<type>` only has `content` property, while `Toml<type>Parser` is responsible for parsing and producing it.</div>

**Comment by Peanuuutz** [comment]
Sync the [answer](https://github.com/akuleshov7/ktoml/discussions/92#discussioncomment-1954075).

New [question](https://github.com/akuleshov7/ktoml/discussions/92#discussioncomment-1954685) raised.

**Comment by akuleshov7** [comment]
> Sync the [answer](https://github.com/akuleshov7/ktoml/discussions/92#discussioncomment-1954075).
> 
> New [question](https://github.com/akuleshov7/ktoml/discussions/92#discussioncomment-1954685) raised.

Yes, let it be open. At least we will mark internal methods with opt-in and revise private methods
